### DISCUSSION TITLE
Sending Pulse from Robots plugin to UR10 robot

### DISCUSSION THREAD

**Discussion by MaryShafiei** [visose/Robots#168] [Q&A] [answered]
Hi everyone. I've been trying to use UR10 robot with an extruder for 3D printing. I've defined direction, enable/disable function, and pulse as digital outputs in Robots plugin but noticed pulse does not work. Then, I tried the 'PulseDO' command in the plugin to send pulse to the extruder but no success as apparently this command does not work with UR robots. Do you have any idea if there is a way to overcome this issue within the plugin? Is there any update coming soon to solve this? 
I've tried to use Arduino to send pulse separately to the extruder while controlling pathways and other parameters in Grasshopper. However, it causes several issues to create a seamless workflow which is crucial in my experiment. I appreciate any advice and help.
Thanks,
Maryam

**Comment by KonradJuenger** [comment]
Hi Maryam, 

I made a script to pulse a digital output on UR. It works by opening a new thread and setting the output to high, waiting, then setting it to low again. 

There are some issues with it: you'll have to filter out pulses with the same length and you'll can overload the controller by sending many pulses 

[PulseDO](https://github.com/KonradJuenger/gh_scripts/blob/8740d1c9868f00f13e1e5cf2fc49bdc3fd87ccc2/pulseDO%20(1).gh)

**Reply by visose** [reply]
Nice, using threads seems like the way to implement the pulseDO command.

But just to clarify first what type of pulsing signal you both need:
The `pluseDO` command in Robots it intended to be used as a 'one-off', as opposed to continuously. Use cases are for example to toggle a solenoid valve (to open or close a gripper), or to indicate to another controller/mechanism that it's their turn to do something. 

It turns on the signal, some fraction of a second later it turns off, and that's it. You might trigger it later again in the program, but not repeatedly in succession.

In @KonradJuenger example, it seems to be pulsing on every target, and in @MaryShafiei, she seems to describe the common stepper motor inputs, where pulse in this case would also be a very fast and continuously pulsing signal. 

In the UR forums there is a similar solution for a continuously pulsed signal, but uses a single thread with an infinite loop, and a boolean to turn it on and off, this will avoid overloading the robot:
https://forum.universal-robots.com/t/making-an-output-pulse-while-moving-the-robot/19002

But for a stepper motor, you should use an external microcontroller for the pulse signal, not the UR robot.

I can implement the pluseDO command in Robots in a similar way to @KonradJuenger example, but this would be for one-off pulsing.

**Reply by KonradJuenger** [reply]
Thanks for the hint and - yes I used pulseDO to communicate to an arduino separated by an optocoupler. For my needs sending one pulse per target worked fine, but setting up persistent thread and only updating the pulse width is much better of course 🙂.

**Reply by MaryShafiei** [reply]
Awesome, thanks @KonradJuenger and @visose for your time and help. 
As @visose mentioned, I need a fast and continuous pulse signal for a stepper motor. It seems the external controller (in my case Arduino) for sending continuous signal should be the option then :-)

**Reply by visose** [reply]
Update 1.5.1 has the PulseDO command working for UR, but it's the 'one-off' version.

The signal is turned on in the main thread, then a second thread runs that switches the signal off after a delay.
The reasoning for this is that this might sync better with robot motion than having it turn on in the second thread, but could as well have no difference. Turning the signal on at the right time is more important than when it turns back off.

**Reply by KonradJuenger** [reply]
This was quick! Thanks for the update and including pulseDO.

**Reply by MaryShafiei** [reply]
Amazing, thanks @visose for the update.

**Comment by visose** [comment, marked as answer]
Closing this as resolved for the original UR PulseDO request. UR `PulseDO` support was added as a one-off pulse command; for fast continuous stepper pulses, use a separate controller or another real-time-capable interface rather than trying to generate every pulse from the robot program.
